### PR TITLE
2325 add integrator variable limits

### PR DESCRIPTION
**Pull request by ErwanGuichard49**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [ ] unit tests and non regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  Types.PerUnit v "Integrator input not accounting for the freeze";
+  Types.PerUnit w(start = Y0) "Non-limited integrator output";
+
+  parameter Types.PerUnit Y0 = 0 "Initial or guess value of output (must be in the limits limitMin .. limitMax)" annotation(
+    Dialog(group="Initialization"));
+
+equation
+  v = K * u;
+
+  der(w) = v * (1 - tanh((w - limitMax + Tol) / Tol)) * (1 - tanh((limitMin + Tol - w) / Tol)) / 4;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:40]: I am not a big fan of the use of this tanh. Could you try to go back to a version where you implement the same block as in Eurostag ? We will then try to figure out what's not working in the test case. Thanks

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  if limitMin > limitMax and DefaultLimitMax then
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:42]: For the case limitMin > limitMax, shouldn't we raise an error ? I believe an assert would be better, no ? How is it done in Eurostag ?

Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:42]: In the Eurostag model, the integrator is followed by a block MIN (with limitMax) then a block MAX (limitMin) or the reverse. 
If limitMin > limitMax :
MIN then MAX leads to limitMin, MAX then MIN leads to limitMax.

I am trying to be as faithful to the Eurostag model as possible.

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  Boolean isFrozen(start = Frozen0) "If true, integration is frozen";
+  Boolean keepFreezing(start = Frozen0) "If true, integration stays frozen";
+  Boolean startFreezing(start = Frozen0) "If true, integration becomes frozen";
+  Types.PerUnit v "Integrator input accounting for the feedback signal but not for the freeze";
+  Types.PerUnit w(start = Y0) "Non-limited integrator output";
+
+equation
+  v = K * u + Kf * (y - w);
+
+  startFreezing = (w < limitMin and v < 0) or (w > limitMax and v > 0);
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:45]: We could add a comment here explaining what is done here

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  keepFreezing = ((w < limitMin + Tol and v < 0) or (w > limitMax - Tol and v > 0)) and pre(isFrozen);
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:46]: same here : the integrator unfreezes only when it is far enough from the limits

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  end if;
+
+  annotation(preferredView = "text",
+  Documentation(info= "<html><head></head><body><p>
+This blocks computes <strong>w</strong> as <em>integral</em>
+of the input <strong>u</strong> multiplied by the gain <em>K</em>.</p>
+
+<p>If the integral reaches a given upper or lower <b>limit</b>, the
+integration is halted and only restarted if the input drives
+the integral away from the bounds.</p>
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:70]: with a sufficient margin

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        Line(points={{-80,78},{-80,-90}}, color={192,192,192}),
+        Polygon(lineColor = {192, 192, 192}, fillColor = {192, 192, 192}, fillPattern = FillPattern.Solid, points = {{-80, 90}, {-88, 68}, {-72, 68}, {-80, 90}}),
+        Line(points={{-90,-80},{82,-80}}, color={192,192,192}),
+        Polygon(lineColor = {192, 192, 192}, fillColor = {192, 192, 192}, fillPattern = FillPattern.Solid, points = {{90, -80}, {68, -72}, {68, -88}, {90, -80}}),
+        Line(points = {{-80, -80}, {20, 20}, {80, 20}}, color = {0, 0, 127}),
+        Text(lineColor = {192, 192, 192}, extent = {{0, -10}, {60, -70}}, textString = "I"),
+        Text(
+          extent={{-150,-150},{150,-110}},
+          textString="K=%K"),
+        Line(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:90]: Last remark, could you rebase and squash ? Thanks

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+model IntegratorVariableLimits "Integrator with limited value of output (variable limits) and freeze"
+  import Modelica;
+  import Dynawo.Types;
+
+  extends Modelica.Blocks.Interfaces.SISO(y(start = Y0));
+
+  parameter Boolean DefaultLimitMax = true "If limitMin > limitMax : if true, y = limitMax, if false, y = limitMin";
+  parameter Types.PerUnit K = 1 "Integrator gain";
+  parameter Types.PerUnit Kf = 0 "Feedback gain";
+  parameter Types.PerUnit Tol = 0 "Tolerance on limit crossing";
```
[the remark this answers is not in the document]
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:25]: Tolerance on limit crossing, aimed at avoid chattering i.e. rapid swings between frozen and unfrozen states

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo (hunk 4 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo:46]: Integration is unfrozen if the integrator output is within the limits and far enough from them or if the integrator input tends to bring the output back within the limits

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo (hunk 3 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo]: ok

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo (hunk 4 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo]: ok

**Review by cossartque — APPROVED** [review]

**Pull request merged**